pkg/tfgen/namecheck.go

```diff
@@ -17,150 +17,341 @@ package tfgen
+		return x.elem.SchemaMap()
+	}
+	return nil, fmt.Errorf("SchemaMap are undefined on an collectionType")
 }
 
-func newNameCollisionError(nc nameContext, tfKey, tfAttr, pulumiName, tfAttrOther string) nameError {
-	m := fmt.Sprintf("Pulumi name %q already represents another Terraform name %q",
-		pulumiName, tfAttrOther)
-	return nameError{context: nc, tfKey: tfKey, tfAttr: tfAttr, message: m}
+func (x *collectionType) Element() (typ, error) {
+	return x.elem, nil
 }
 
-func validatePropertyNamesForProvider(prov tfbridge.ProviderInfo) nameErrors {
-	errors := []nameError{}
+var _ typ = (*collectionType)(nil)
 
-	prov.P.Schema().Range(func(tfConfigKey string, sch shim.Schema) bool {
-		es := validatePropertyNames(configPropertyNames, tfConfigKey, prov.P.Schema(), prov.Config)
-		errors = append(errors, es...)
-		return true
-	})
+type scalarType struct{}
 
-	prov.P.ResourcesMap().Range(func(tfResKey string, res shim.Resource) bool {
-		info := prov.Resources[tfResKey]
-		if info == nil {
-			info = &tfbridge.ResourceInfo{}
-		}
-		res.Schema()
-		es := validatePropertyNames(resourcePropertyNames, tfResKey, res.Schema(), info.Fields)
-		errors = append(errors, es...)
-		return true
-	})
-
-	prov.P.DataSourcesMap().Range(func(tfDsKey string, ds shim.Resource) bool {
-		info := prov.DataSources[tfDsKey]
-		if info == nil {
-			info = &tfbridge.DataSourceInfo{}
-		}
-		es := validatePropertyNames(dataSourcePropertyNames, tfDsKey, ds.Schema(), info.Fields)
-		errors = append(errors, es...)
-		return true
-	})
-
-	return nameErrors(errors)
-}
-
-func validatePropertyNames(
-	nc nameContext,
-	tfKey string,
-	schemaMap shim.SchemaMap,
-	infos map[string]*tfbridge.SchemaInfo,
-) []nameError {
-	errors := []nameError{}
-	mapping := map[string]string{}
-	schemaMap.Range(func(tfAttr string, tfAttrSchema shim.Schema) bool {
-		pulumiName := tfbridge.TerraformToPulumiNameV2(tfAttr, schemaMap, infos)
-
-		if oldAttr, exists := mapping[pulumiName]; exists {
-			errors = append(errors, newNameCollisionError(nc, tfKey, tfAttr, pulumiName, oldAttr))
-		}
+func (x *scalarType) Element() (typ, error) {
+	return nil, fmt.Errorf("Element undefined on an scalarType")
+}
 
-		mapping[pulumiName] = tfAttr
+func (x *scalarType) Property(pn paths.PropertyName) (typ, error) {
+	return nil, fmt.Errorf("Properties are undefined on an scalarType")
+}
 
-		tfAttrRecovered := tfbridge.PulumiToTerraformName(pulumiName, schemaMap, infos)
-		if tfAttrRecovered != tfAttr {
-			errors = append(errors, newNameTurnaroundError(nc, tfKey, tfAttr, pulumiName, tfAttrRecovered))
-		}
+func (x *scalarType) SchemaMap() (shim.SchemaMap, error) {
+	return nil, fmt.Errorf("Fields are undefined on an scalarType")
+}
+
+func (x *scalarType) SchemaInfos() (map[string]*tfbridge.SchemaInfo, error) {
+	return nil, fmt.Errorf("Fields are undefined on an scalarType")
+}
+
+var _ typ = (*scalarType)(nil)
 
-		return true
+func parseSchema(s shim.Schema, info *tfbridge.SchemaInfo) typ {
```

Comment:
`parseSchema` is not a clear name for what this function is doing. It looks like it is converting a schema + info into a `typ`. Would it make sense to call it `newTypFromSchemaAndInfo`?

Reply:
This is some 1am code. `tip` is not a clear name either. It's all internal helpers to make sense out of pretty crazy type encoding that is tfbridge.SchemaInfo so that the code here can step down and lookup matching SchemaInfo by TypePath.

Reply:
Let me think about this a bit, how to name this.